instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/SnsImpl.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.awssdk.v2_2;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapPropagator;
+import java.util.HashMap;
+import java.util.Map;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.services.sns.SnsClient;
+import software.amazon.awssdk.services.sns.model.MessageAttributeValue;
+import software.amazon.awssdk.services.sns.model.PublishRequest;
+
+// this class is only used from SnsAccess from method with @NoMuzzle annotation
+class SnsImpl {
+  static {
+    // Force loading of SnsClient; this ensures that an exception is thrown at this point when the
+    // SNS library is not present, which will cause SnsAccess to have enabled=false in library mode.
+    @SuppressWarnings("unused")
+    String ensureLoadedDummy = SnsClient.class.getName();
+  }
+
+  private SnsImpl() {}
+
+  static SdkRequest modifyRequest(
+      SdkRequest request, Context otelContext, TextMapPropagator messagingPropagator) {
+    if (messagingPropagator == null) {
+      return null;
+    } else if (request instanceof PublishRequest) {
+      return injectIntoPublishRequest((PublishRequest) request, otelContext, messagingPropagator);
+    } else {
+      // NB: We do not support PublishBatchRequest which was only introduced in 2.17.84.
+      // To add support, either raise the minimum requirement for the SNS-specific features, or use
+      // reflection + NoMuzzle (NB: Can use SnsRequest base class, MessageAttributeValue class,
```

Comment:
I think that raising the minimum compileOnly version should be fine as long as we have tests that cover 2.2, and verify that this works even without `PublishBatchRequest` being present.

Reply:
Right, one would only need to sprinkle a few (hopefully only one) `@NoMuzzle` so that the other code is still loaded if an older version is used at runtime, right?

Reply:
Yeah -- `@NoMuzzle` and perhaps a `Class.forName()` check somewhere would probably suffice.

Reply:
I think I won't do this in this PR since it is self-contained follow-up work, but in case we want to do it in the future, I captured the response of a PublishBatch request using two messages with entry IDs i1 and i2 for use in the unit tests:

```xml
<PublishBatchResponse xmlns="http://sns.amazonaws.com/doc/2010-03-31/">
  <PublishBatchResult>
    <Failed/>
    <Successful>
      <member>
        <MessageId>51416048-2afd-54c9-be4f-3d50f9062398</MessageId>
        <Id>i1</Id>
      </member>
      <member>
        <MessageId>1a2260b6-2e62-5192-a623-3daecd99b6c4</MessageId>
        <Id>i2</Id>
      </member>
    </Successful>
  </PublishBatchResult>
  <ResponseMetadata>
    <RequestId>733a6d44-b19e-5c97-8995-305c52654bd6</RequestId>
  </ResponseMetadata>
</PublishBatchResponse>
```

(There was no example at https://docs.aws.amazon.com/sns/latest/api/API_PublishBatch.html)

Reply:
> I think I won't do this in this PR since it is self-contained follow-up work,

Definitely 💯 
This PR is fine as it is, and small PRs are a delight to review -- I'd prefer the SNS batch thing to be a separate one if you still want to do that in the future.

Reply:
I adapted the comment here in 722c0f39193396b3ef11a1ff64108af8141da7f1 though, to not lose this thread in case I, or anybody else wants to tackle it.